internal/cephfs/groupcontrollerserver.go

```diff
@@ -0,0 +1,756 @@
+			if err != nil {
+				log.ErrorLog(ctx, "failed to delete snapshot and undo reservation: %v", err)
+			}
+		}
+	}()
+
+	responses := make([]csi.CreateSnapshotResponse, 0)
+	for _, volID := range req.GetSourceVolumeIds() {
+		if !freeze {
+			err = fsQuiesceWithExpireTimeout(ctx, vgo.RequestName, fsMap)
+			if err != nil {
+				log.ErrorLog(ctx, "failed to quiesce filesystem with timeout: %v", err)
+
+				return nil, err
+			}
+		}
+		// Create the snapshot for the volumeID
+		clusterID := getClusterIDForVolumeID(fsMap, volID)
+		if clusterID == "" {
+			return nil, fmt.Errorf("failed to get clusterID for volumeID %s", volID)
+		}
+		req := &csi.CreateSnapshotRequest{
+			SourceVolumeId: volID,
+			Name:           vgs.FsVolumeGroupSnapshotName + "-" + volID,
+			Secrets:        secrets,
+			Parameters: map[string]string{
+				"clusterID": clusterID,
+			},
+		}
+
+		resp, err = cs.createSnapshotAndAddMapping(ctx, req, vgo, vgs)
+		if err != nil {
+			// Handle cleanup
+			log.ErrorLog(ctx, "failed to create snapshot: %v", err)
+
+			return nil, err
+		}
+		responses = append(responses, *resp)
+	}
+
+	if !freeze {
+		err = releaseFSQuiesce(ctx, vgo.RequestName, fsMap)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to release filesystem quiesce: %v", err)
+
+			return nil, err
+		}
+	}
+
+	return &responses, nil
+}
+
+// releaseSubvolumeQuiesce releases the quiesce of the subvolumes and subvolume
+// groups in the filesystems for the volumeID's present in the
+// CreateVolumeGroupSnapshotRequest.
+func releaseFSQuiesce(ctx context.Context,
+	requestName string,
+	fsMap map[string]core.FSQuiesceClient,
+) error {
+	inProgress := false
+	var err error
+	var data *admin.QuiesceInfo
+	for _, fm := range fsMap {
+		// UnFreeze the filesystems, subvolumes and subvolume groups
+		data, err = fm.ReleaseFSQuiesce(ctx, requestName)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to release filesystem quiesce: %v", err)
+
+			return err
+		}
+		state := core.GetQuiesceState(data.State)
+		if state != core.Released {
+			inProgress = true
+
+			continue
+		}
+	}
+
+	if inProgress {
+		return cerrors.ErrQuiesceInProgress
+	}
+
+	return nil
+}
+
+// fsQuiesceWithExpireTimeout quiesces the subvolumes and subvolume
+// groups in the filesystems for the volumeID's present in the
+// CreateVolumeGroupSnapshotRequest.
+func fsQuiesceWithExpireTimeout(ctx context.Context,
+	requestName string,
+	fsMap map[string]core.FSQuiesceClient,
+) error {
+	var err error
+
+	var data *admin.QuiesceInfo
+	inProgress := false
+	for _, fm := range fsMap {
+		// reinitialize the expiry timer for the quiesce
+		data, err = fm.FSQuiesceWithExpireTimeout(ctx, requestName)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to quiesce filesystem with timeout: %v", err)
+
+			return err
+		}
+		state := core.GetQuiesceState(data.State)
+		if state == core.Quiescing {
+			inProgress = true
+
+			continue
+		}
+		if state != core.Quiesced {
```

Comment:
```suggestion
		} else if state != core.Quiesced {
```

I think this makes it simpler. 
Let's go with this only if it makes sense to you too.